tests/DCPS/XTypes/XTypes.cpp

```diff
@@ -0,0 +1,667 @@
+
+  PlainCdrStruct pcs;
+  pcs.key = key_value;
+  pcs.value = 1;
+  pdw->write(pcs, HANDLE_NIL);
+  if (verbose) {
+    ACE_DEBUG((LM_DEBUG, "writer: PlainCdrStruct\n"));
+  }
+}
+
+
+void write_final_struct(const DataWriter_var& dw)
+{
+  FinalStructDataWriter_var pdw = FinalStructDataWriter::_narrow(dw);
+
+  FinalStruct fs;
+  fs.key = key_value;
+  pdw->write(fs, HANDLE_NIL);
+  if (verbose) {
+    ACE_DEBUG((LM_DEBUG, "writer: FinalStruct\n"));
+  }
+}
+
+
+void write_modified_final_struct(const DataWriter_var& dw)
+{
+  ModifiedFinalStructDataWriter_var pdw = ModifiedFinalStructDataWriter::_narrow(dw);
+
+  ModifiedFinalStruct mfs;
+  mfs.key = key_value;
+  mfs.additional_field = FINAL_STRUCT_AF;
+  pdw->write(mfs, HANDLE_NIL);
+  if (verbose) {
+    ACE_DEBUG((LM_DEBUG, "writer: ModifiedFinalStruct\n"));
+  }
+}
+
+
+void write_appendable_struct(const DataWriter_var& dw)
+{
+  AppendableStructDataWriter_var typed_dw = AppendableStructDataWriter::_narrow(dw);
+
+  AppendableStruct as;
+  as.key = key_value;
+  typed_dw->write(as, HANDLE_NIL);
+  if (verbose) {
+    ACE_DEBUG((LM_DEBUG, "writer: AppendableStruct\n"));
+  }
+}
+
+
+void write_additional_prefix_field_struct(const DataWriter_var& dw)
+{
+  AdditionalPrefixFieldStructDataWriter_var typed_dw = AdditionalPrefixFieldStructDataWriter::_narrow(dw);
+
+  AdditionalPrefixFieldStruct apfs;
+  apfs.key = key_value;
+  apfs.additional_field = APPENDABLE_STRUCT_AF;
+  typed_dw->write(apfs, HANDLE_NIL);
+  if (verbose) {
+    ACE_DEBUG((LM_DEBUG, "writer: AdditionalPrefixFieldStruct\n"));
+  }
+}
+
+
+void write_additional_postfix_field_struct(const DataWriter_var& dw)
+{
+  AdditionalPostfixFieldStructDataWriter_var typed_dw = AdditionalPostfixFieldStructDataWriter::_narrow(dw);
+
+  AdditionalPostfixFieldStruct apfs;
+  apfs.key = key_value;
+  apfs.additional_field = APPENDABLE_STRUCT_AF;
+  typed_dw->write(apfs, HANDLE_NIL);
+  if (verbose) {
+    ACE_DEBUG((LM_DEBUG, "writer: AdditionalPostfixFieldStruct\n"));
+  }
+}
+
+
+void write_mutable_struct(const DataWriter_var& dw)
+{
+  MutableStructDataWriter_var typed_dw = MutableStructDataWriter::_narrow(dw);
+
+  MutableStruct ms;
+  ms.key = key_value;
+  ms.additional_field = MUTABLE_STRUCT_AF;
+  typed_dw->write(ms, HANDLE_NIL);
+  if (verbose) {
+    ACE_DEBUG((LM_DEBUG, "writer: MutableStruct\n"));
+  }
+}
+
+
+void write_modified_mutable_struct(const DataWriter_var& dw)
+{
+  ModifiedMutableStructDataWriter_var typed_dw = ModifiedMutableStructDataWriter::_narrow(dw);
+
+  ModifiedMutableStruct ams;
+  ams.key = key_value;
+  ams.additional_field = MUTABLE_STRUCT_AF;
+  typed_dw->write(ams, HANDLE_NIL);
+  if (verbose) {
+    ACE_DEBUG((LM_DEBUG, "writer: ModifiedMutableStruct\n"));
+  }
+}
+
+
+void write_mutable_union(const DataWriter_var& dw)
+{
+  MutableUnionDataWriter_var typed_dw = MutableUnionDataWriter::_narrow(dw);
+
+  MutableUnion mu;
+  mu.key(key_value);
```

Comment:
Set the discriminator too?

Reply:
The methods in the union that are named for its members always set the discriminator.

Reply:
That's nice.